gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java

```diff
@@ -103,72 +118,165 @@ public StructType getSchema() {
+                "SELECT %s FROM %s WHERE %s = (SELECT MAX(%s) FROM %s WHERE %s = '%s') LIMIT 1",
                 COL_GC_RUN_ID,
                 //
                 catalogAndTableWithRefName,
                 //
                 COL_GC_RUN_START,
                 COL_GC_RUN_START,
-                catalogAndTableWithRefName)
+                catalogAndTableWithRefName,
+                //
+                COL_ROW_TYPE,
+                RowType.CONTENT_OUTPUT.name())
             .collectAsList();
     return rows.isEmpty() ? Optional.empty() : Optional.of(rows.get(0).getString(0));
   }
 
+  public Row createCheckPointMarkerRow(String runId) {
+    return RowFactory.create(
+        getGcStartTime(runId),
+        runId,
+        RowType.CHECKPOINT_MARKER.name(),
+        null,
+        null,
+        null,
+        null,
+        null,
+        null,
+        null,
+        null);
+  }
+
+  public void writeToOutputTable(List<Row> rows) {
+    writeToOutputTable(sparkSession.createDataFrame(rows, schema));
+  }
+
   void writeToOutputTable(Dataset<Row> rowDataset) {
     try {
       // write content rows to the output table
       rowDataset.writeTo(catalogAndTableWithRefName).append();
     } catch (NoSuchTableException e) {
       throw new RuntimeException(
-          "Problem while writing gc output rows to the table: " + catalogAndTableWithRefName, e);
+          "Problem while writing output rows to the table: " + catalogAndTableWithRefName, e);
     }
   }
 
   static Row createContentRow(
-      Content content, String runId, Timestamp startedAt, long snapshotId, Reference ref) {
+      Content content,
+      String runId,
+      Timestamp startedAt,
+      long snapshotId,
+      Reference ref,
+      String commitHash,
+      String metadataLocation,
+      boolean isExpired) {
     return RowFactory.create(
         startedAt,
         runId,
+        RowType.CONTENT_OUTPUT.name(),
         content.getId(),
         content.getType().name(),
         snapshotId,
         ref.getName(),
-        ref.getHash());
+        ref.getHash(),
+        commitHash,
+        metadataLocation,
+        isExpired);
   }
 
-  private void createTableIfAbsent(
+  static Row createCheckPointRow(String runId, Timestamp startedAt, LiveContentsResult result) {
+    return RowFactory.create(
+        startedAt,
+        runId,
+        RowType.CHECKPOINT.name(),
+        null,
+        null,
+        null,
+        result.getReferenceName(),
+        result.getHashOnReference(),
+        result.getLastLiveCommitHash(),
+        null,
+        null);
+  }
+
+  /**
+   * Finds the latest completed checkpoint. Checkpoint is considered as completed when it has a
+   * checkpoint marker row (written when the expiry operation is successful).
+   *
+   * @return a map of referenceName and checkpoint (its last traversed live commit hash from the
+   *     head).
+   */
+  Map<String, String> collectLatestCommitCheckPoint() {
```

Comment:
What happens to this checkpoint information, if this happens:

BEFORE:
* branch A @ commit 123
* branch B @ commit 456

THEN: GC runs - checkpoints written

THEN: User assigns B to 123 and A to 456

The checkpoint information will be wrong, no?